### DISCUSSION TITLE
Vue.js, vale a pena?

### DISCUSSION THREAD

**Discussion by mathvbarone** [frontendbr/forum#732] [Perguntas e Respostas]
Galera, eu ainda nao peguei pra estudar a fundo nenhum framework, fiz um curso de React na Udemy mas foi bem basico. 

To vendo que muita gente ta amando o Vue.js. Alguem saberia pontuar o que o Vue.js tem de tao especial que os outros nao tem? E alem disso, vale a pena focar nele, visto que a maioria esmagadora de vagas no mercado é pra Angular/React?

**Comment by expalmer** [comment]
Vale a penas estudar Javascript!
React, Vue, Angular e os demais frames fazem a mesma coisa de maneiras diferentes. 
Estude os patterns do Javascript que vai conseguir usar e entender qualquer framework.
Aconselho a entrar no core de algum framework e ver como ele funciona, como foi escrito, como ele resolve determinado problema...
Esse aqui é bem legal para entender algumas coisinhas [Choo](https://github.com/choojs/choo)

**Comment by thulioph** [comment]
Além do que foi dito acima..

Se você gosta de se manter atualizado e ir um pouco mais além, eu aconselharia você a fazer pequenos experimentos com cada framework e daí tirar a sua conclusão.

> Não vá pela maioria!

Existe todo aquele papo de que depende do seu problema, mas ele é bem legal sim, vale ao menos dar uma olhada, respirar um pouco daquele ar, pra ir compreendendo. Tem muito material brasileiro na internet, a documentação é uma das melhores e todo o ecossistema é tranquilo de entender, sem falar na comunidade que é bem movimentada. 

- [Documentação](https://vuejs.org/v2/guide/)
- [Artigos](http://www.vuejs-brasil.com.br/)
- [Slack](slack.vuejs-brasil.com.br)

Caso sirva de ajuda, eu compartilhei um pouco do que sei em alguns encontros e os slides você pode conferir [aqui](https://speakerdeck.com/thulioph/vuejs). 

Espero ter ajudado 😃

**Comment by mathvbarone** [comment]
Obrigado pelas respostas pessoa, concordo com vocês em todos os pontos. @expalmer você poderia ser um pouco mais especifico em relação a esses patterns js?

**Comment by caioincau** [comment]
Vale a pena sim, sou meio suspeito já que lancei o livro de Vue.js pela Casa do Código recentemente rsrs

Bom, eu acredito que a curva de aprendizado do Vue.js é menor do que a dos seus concorrentes como o React e o Angular, ainda tem a vantagem de performance, ele é mais leve e no geral, mais rápido.
https://br.vuejs.org/v2/guide/comparison.html

Além de ser MIT License, ao contrário do React que é BSD e tem patentes(https://github.com/frontendbr/forum/issues/670).

**Comment by caioincau** [comment]
Sobre Patterns, acho que ele estava se referindo a Design Patterns, que são (a grosso modo), meios de solução de problemas comum, que foram padronizadas.

https://addyosmani.com/resources/essentialjsdesignpatterns/book/

**Comment by mathvbarone** [comment]
Obrigado pela ajuda @icaioincau!

**Comment by Cleysonlb** [comment]
Ainda estou com essa duvida também, mas peguei pra aprender mesmo assim e estou achando bem legal, muito divertido e interessante, independente se vale a pena ou não, vai la e aprende... Os conceitos que você aprender não serão atoa! O Vue usa html, css e JS puro ou seja, oq vc aprender lá pode ser aplicado fora também 💃

**Comment by heliomsolivas** [comment]
Eu resolvi colocar em um projeto aqui sem nunca ter usado e realmente a curva de aprendizado é menor (acredito eu), acho que vale a pena sim e cada vez tenho visto mais gente falar de Vue.js

**Comment by expalmer** [comment]
Como o @icaioincau disse, é isso mesmo!
Tem o antigo e famoso [http://jstherightway.org/pt-br/](http://jstherightway.org/pt-br/).
Eu aprendi muito lendo esse do [do Addy Osmani](https://addyosmani.com/resources/essentialjsdesignpatterns/book/) que o @icaioincau comentou, e também o [Javascript Patterns do Stoyan Stefanov](https://www.amazon.com/JavaScript-Patterns-Better-Applications-Coding/dp/0596806752)
E concordo com a opinião da galera aí, o negócio é experimentar. Não fica com aquele pensamento __"Poxa escolhi o framework X e todo mundo tá falando do Y, estou ficando para trás"__...não rapaz, se você sabe Javascript, você usa um framework e começa a identificar que o framework X usa tal e tal coisa, o framework Y usa isso e aquilo...
E segue perguntando e tirando dúvidas com a galera da comunidade, tem muito fera aqui, eu aprendo muito com eles.

**Comment by danilovaz** [comment]
@mathvbarone Falando em vagas, na minha opinião não acredito que tenha que ser um fator tão importante, apesar de sim ter que ser um fator. Se pensarmos em 2009/2010, o Ruby/Ruby on Rails ainda estava ganhando tração de mercado no Brasil, haviam mais vagas de quê? Java 😄  Se olhássemos somente vagas, não teríamos saído de Cobol -> Java -> PHP -> Ruby (meta frameworks) -> JS (frameworks) etc, etc. As vezes temos que experimentar coisas fora do mercado e usar o que resolve o problema em questão.

> `Alguem saberia pontuar o que o Vue.js tem de tao especial que os outros nao tem?`

Isso é fácil, o Vue.js veio depois de todos os outros grandes frameworks (ou libs + ecossistema) e pôde aprender com os erros dos outros, sendo assim ele já começou melhor que os outros que tiveram que aprender no meio do caminho. Só para contextualizar, os lançamentos:
- AngularJS: 2010
- Ember.js: 2011
- React: 2013
- Vue.js: 2014
- Angular (2.x): 2016

Isso não é nenhum demérito, é apenas fato, a ordem cronológica nos mostra também a evolução das arquiteturas de front: AngularJS > Ember.js > React > Vue.js

*: Angular 2.x é um caso a parte, já que ele é na lista o framework mais novo (não levando em consideração o AngularJS, já que o code base foi totalmente reescrito), deixo para alguém mais focando nele dar uma opinião mais formada.

Dos Frameworks JS em atuação hoje, o [Ember.js](https://emberjs.com/) é mais antigo ainda em atuação e ainda moderno. Diferente do AngularJS que fez um outro framework e reescreveu seu code base, no Ember foi possível evoluir da API 1.x para a 2.x sem quebrar projeto de ninguém e melhorar em muito a performance.

Alguns pontos da comparação do Vue vs Ember, que estão [nesse link](https://vuejs.org/v2/guide/comparison.html) que eu queria comentar, seguindo a ordem em que elas são citadas lá:

1 - Das comparações, é a menor. E não sei bem o motivo, uma vez que tanto a forma de Router (vue-router) quanto o Persistência de Dados (vuex) são claramentes uma evolução inspiradas em Ember Router e Ember Data, respectivamente. Isso, claro, não dá para comparar Ember e Vue sem levar em conta Vuex, assim como não dá para comparar Ember e React sem levar em consideração todo o ecossistema.

2 - Quem diz que o Handlebars limita, está um tanto quanto equivocado. Eu consigo usar todo o poder do Handlebars quanto estender e utilizar da forma que preciso. Acontece que no Ember, já temos vários bult-in helpers que não há necessidade de se criar alguma expressão e se houver, é fácil de fazê-lo e até tornar isso reutilizável em outros projetos.

3 - O Ember 1.13 (2014) era uma tragédia em performance, realmente. Mas testes de benchmark não é tudo, quando o assunto é performance. Porque um framework pode ser performático em uma determinada tarefa em determinado cenário, mas péssimo com cenários com outras variantes. Contudo, é um ponto. O teste de benchmark citado na comparação, a versão que comparam do Ember é a 2.13 (Abril de 2017), de lá para cá houveram mais 2 releases: [2.14](https://www.emberjs.com/blog/2017/07/06/ember-2-14-released.html) e [2.15](https://www.emberjs.com/blog/2017/09/01/ember-2-15-released.html)

![initial-render-698300e7](https://user-images.githubusercontent.com/1218337/30645124-206064e4-9deb-11e7-9662-989c125489b9.png)

E a cada release o time está melhorando a performance, importante citar que de 2016 até agora eles focaram mais em melhorar a arquitetura do Ember, para justamente melhorar pontos citados nessas comparações que são verdade (ainda), o core team mapeou com a comunidade alguns pontos:

![ember-hierarchy-of-skepticism-4673d574](https://user-images.githubusercontent.com/1218337/30645266-8e93f7e6-9deb-11e7-8a1a-01b77cbdea9f.png)

Os pontos 1 e 2 não são verdades mais, uma vez que a documentação do Ember é excelente e não existe mais bind com o helper `{{bindAttr}}`, é tudo Data Down, Actions Up (ou Data Down, Event Up).

O File Size do Ember ainda estão em busca de melhorá-lo mais e mais, contudo o Glimmer.js não deixa nada a desejar:

![hello-world-cli-size-677833c7 1](https://user-images.githubusercontent.com/1218337/30646114-42065e2a-9dee-11e7-89d2-fd14348398d0.png)

Com isso, estão focando agora no Ember em mudar de um Objeto Ember para Classes ES6, então será possível escrever o bom e velho plain javascript, ao invés de amarrar em um obejto Ember 😄 

Ember será modularizado, adeus afirmações de que o Ember é monolíto. Assim, você conseguirá começar um projeto com a UI Engine do Ember (Glimmer.js) e conforme (e se) sua aplicação for ficando mais complexa e você precisar de: Router, Services, Data etc, você conseguirá inserir esses módulos.

![spectrum-of-app-complexity-5482cee7 1](https://user-images.githubusercontent.com/1218337/30645499-4c0ef5b4-9dec-11e7-9d09-b228146ac9be.png)

Hoje a UI Engine: [Glimmer.js](https://glimmerjs.com/), é tão rápida quanto o Preact tanto no Initial Render, quanto no Re-Render, você pode conferir [aqui](http://yehudakatz.com/2017/04/05/the-glimmer-vm-boots-fast-and-stays-fast/). E como eu disse, o time do Ember *ainda* nem focou suas atenções diretamente para performance.

Aí fica a dúvida: mas e a curva de aprendizagem? A curva de aprendizagem do Ember é mais alta para quem não entende o JavaScript (closure, context, scope, etc), e como o @expalmer bem colocou: aprenda bem o JavaScript. 

Porque, na minha opinião, qualquer um desses frameworks/libs vão entregar o que você precisa no final, todos são muito bons. Não existe framework que fique performático quando o código escrito nele é ruim, não tem padrão (design patterns), há complexidade ciclomática alta, etc. Existe muito no desenvolvimento de software que deve ser aprendido pelo Front-End também, principalmente escrevendo JS.

Fica a dica de melhor curso pra aprender bem JS: https://www.udemy.com/curso-javascript-ninja/ e desculpe se fui invasivo citando o Ember, mas como colocaram o link de comparação e ninguém citou e como eu uso há 2 anos, quis deixar meus 20 centavos e espero ter sido útil.

E sendo mais direto: Sim, vale a pena estudar e usar o Vue 😄

**Comment by ghost** [comment]
@danilovaz por favor transforme isso em artigo. Eu realmente compartilho da mesma idéia =)

**Comment by felipepalazzo** [comment]
Sobre se _vale a pena_ ou não, acho que o primeiro passo é se perguntar: que objetivo quero atingir ao investir meu tempo aprendendo essa lib e/ou framework? Melhorar o CV? Iniciar um projeto pessoal? Escrever um artigo? Just have some fun? :upside_down_face: 

Além das vagas, como mencionou, existem outros pontos importantes que também devem ser considerados: tamanho da comunidade, ecossistema, meetups, eventos, livros, cursos, número de perguntas no Stack Overflow, etc.  Pra ficar num único exemplo:

![react_vue_angular](https://user-images.githubusercontent.com/2220082/30656823-bc445364-9e0b-11e7-8d68-78d789effaad.png)
Fonte: [NPM Trends](http://www.npmtrends.com/@angular/core-vs-react-vs-vue)

**Comment by mathvbarone** [comment]
Galera, excelentes pontos abordados, especialmente o @danilovaz  que fez praticamente um artigo, excelente, muito obrigado!

Em relação à questão das vagas, eu não levo isso tão à ferro e fogo, serve apenas como um filtro daquilo que é hype, do que é tecnologia relevante.

Já estou guardando todos os links enviados para estudo e ja comprei um curso de Vue.js na udemy para ao menos aprender o básico do framework :)

Mais uma vez obrigado a todos pela explicação, espero que ajude outras pessoas que também estão com essa dúvida 👍

**Comment by vinicius73** [comment]
Apenas pontuando a resposta do @felipepalazzo 
Infelizmente não é possível usar a métrica de downloads do NPM como referencia no Vue, existe muita gente que utiliza ele via CDN, boa parte do ecossistema Vue é compativel com CDN, e isso não entra nas métricas.
Sem contar com o proxy que existe na ásia/china que também não entra nestes gráficos.

Gosto das métricas do https://bestof.js.org
![image](https://user-images.githubusercontent.com/1561347/30931542-5ef10208-a39b-11e7-9a62-d7e44e3b673e.png)

Dão uma visão geral do cenário.
Recentemente a comunidade Vue.js Brasil criou um repo para quem as empresas que usam Vue no brasil sejam listadas, https://github.com/vuejs-br/empresas-que-usam-vue-no-brasil o crescimento dele é inegável e incrível.

 Mas faço minhas as palavras dos amigos aqui, teste todos e tire você mesmo suas conclusões, acima de tudo, **aprenda JavaScript**
Querendo mais informações da uma olhada nos meus artigos, faço algumas comparações técnicas e não técnicas sobre o assunto.

PS: @danilovaz se você não criar um artigo com esses dados juro que vou te xingar no Twitter.

**Comment by didiraja** [comment]
@vinicius73 tem como passar link dos seus aartigos pra eu dar uma olhada?

**Comment by vinicius73** [comment]
Claro @didiraja 
- https://blog.codecasts.com.br/por-que-vue-js-e-nao-react-d5b58c09d193
- https://blog.codecasts.com.br/vue-js-nao-e-um-angular-simplificado-6394c18cc689
- https://medium.com/by-vinicius-reis/o-que-e-react-ng2-auleria-vue-e34b0c77b5a1

**Comment by garbinmarcelo** [comment]
Estou pesquisando sobre Vue.js e vim parar aqui... atualmente utilizo Laravel em meus projetos na parte de back-end e Bootstrap na parte de front-end. Sendo que com o Laravel eu já trato cada parte do sistema por módulos, sendo assim tenho eles como meus "componentes".

Gostaria de saber o que o Vue.js traria a mais na minha aplicação? qual o beneficio de utilizá-lo? eu teria trabalho em dobro ao ter que criar componentes novos?

Valeuu

**Comment by felipefialho** [comment]
Invoco o @vedovelli

**Comment by raisiqueira** [comment]
Utilizo VueJS + Vuex + WP-API em produção e não tenho do que reclamar, a curva de aprendizado é mais fácil do que os demais, sem falar do arquivo `.vue` que qualquer desenvolvedor que olhe para ele ssaiba oq tal arquivo faz no escopo do projeto. Optei por VueJS pelo rápido setup de um projeto (mesmo o React tendo o `create-react-app`).

**Comment by willycamargo** [comment]
Para complementar:

* https://devnaestrada.com.br/2018/01/05/vale-pena-vuejs.html

**Comment by danilovaz** [comment]
@vinicius73 @edmolima  Demorou mas saiu hahaha -> https://emberbrasil.com.br/o-ember-js-ainda-existe-4e02e2435695
